### ISSUE TITLE
Preprocessor to support `.html.erb` files (Ruby on Rails, ERB)

### ISSUE THREAD

**Issue by Splines** [yeonjuan/html-eslint#223]
Hey there, thanks so much for creating this tool. I'm wondering how all the people programmed HTML all those years without a proper HTML linter? Like as if the [W3C Validator](https://validator.w3.org/) is one of the top results when I search for "HTML linter" on the web 😂

---

I'd like to lint the HTML part of `.html.erb` files. ERB is a templating language used in Ruby on Rails projects. In order to lint `.js.erb` files, I've written a custom [`ESLint plugin`](https://github.com/Splines/eslint-plugin-erb/), which is essentially a [preprocessor](https://github.com/Splines/eslint-plugin-erb/blob/baa20350239835528cd4a7d74f1a0f1432766e37/lib/preprocess.js#L15-L24) (and then also a postprocessor) to strip away all the ERB parts in the code (replacing them with a dummy comment), while keeping track of the positions to correctly map line and colon numbers of ESLint messages. Here's what you can achieve with it (image should be animated):

![293907467-623d6007-b4f5-41ce-be76-5bc0208ed636](https://github.com/user-attachments/assets/5b6447e9-ff41-4bfd-a991-1e4ed82755fe)

I was wondering if your project also supports some kind of mechanism such that I could use this preprocessor again (maybe with only slight changes) for `.html.erb` files. The recommended ESLint configuration basically just [declares the processor](https://github.com/Splines/eslint-plugin-erb/blob/baa20350239835528cd4a7d74f1a0f1432766e37/lib/index.js#L17-L20).

This would allow me to lint the HTML part of these files. I would then use a tool like [`ERB Lint`](https://github.com/Shopify/erb_lint) to lint the ERB parts.

**Comment by yeonjuan** [comment]
@Splines hello. Thanks for the suggestion, I agree with adding a preprocessor.  However, I am not very familiar with ERB, so if I were to implement it, it might take some time. I will refer to the code you shared. Thank you.

**Comment by Splines** [comment]
Hey, I noticed that I didn't have too much work to do to support `html-eslint` with my processor. So I'm happy to announce that with [my new release 2.1.0](https://github.com/Splines/eslint-plugin-erb/releases/tag/v2.1.0), you can now also lint `.html.erb` files with it 🙌

![eslint-html-erb](https://github.com/user-attachments/assets/d0a9942b-646e-4644-808d-9276b81a195a)

**Comment by Splines** [comment]
@yeonjuan I recently stumbled on this case where my plugin currently fails to provide reasonable output. Assume this HTML code (referred to as [1]):

```html
<button id="nice"
  <% if current_user.admin? %>
  class="btn-admin"
  <% end %>
</button>
```

What my plugin gives to `html-eslint` is then (we denote this by [2]):
```html
<button id="nice"
  <!-- 566513c5d83ac26e15414f2754 0 -->
  class="btn-admin"
  <!-- 566513c5d83ac26e15414f2754 1 -->
</button>
```

Linting this with the following setting:

```js
"@html-eslint/sort-attrs": "error",
```

will produce invalid code. I cannot expect you to lint [2] properly, since it isn't even valid HTML: in an attribute list, no comments are allowed. Would it still be possible to tailor for this special case? Or do you have a better idea for what I could use as replacement string?

**Comment by yeonjuan** [comment]
@Splines This is probably a similar issue to https://github.com/yeonjuan/html-eslint/issues/231.
I realize that in the real world it's rare to use purely HTML, as in the example you wrote. There are many different template engines that work with HTML.
I've been thinking about how to support many different kinds of template engines. And I'm thinking that perhaps we should handle this at the [parser](https://github.com/yeonjuan/es-html-parser) level, like we did for JavaScript Template Literal. 
I plan to look at this in February. I don't have a clear direction yet, so if you have any ideas, please share. Thanks!

**Comment by yeonjuan** [comment]
Hi, @@Splines 
I've been working on covering the template engine syntax, and I'm happy to announce that development is almost complete and an alpha version has been released.

```console
npm install -D @html-eslint/eslint-plugin@0.34.0-alpha.0 @html-eslint/parser@0.34.0-alpha.0
```

```js
import eslintHTML from "@html-eslint/eslint-plugin";
import eslintHTMLParser, { TEMPLATE_ENGINE_SYNTAX } from "@html-eslint/parser";

export default [

  {
    "files": ["**/*.html"],
    "languageOptions": {
      parser: eslintHTMLParser,
      parserOptions: {
        templateEngineSyntax: TEMPLATE_ENGINE_SYNTAX.TWIG // Configuration for TWIG
      }
    },
    "plugins": {
      "@html-eslint": eslintHTML,
    },
    "rules": {
     "@html-eslint/sort-attrs": ["error"]
    }
},
]
```

**Comment by Splines** [comment]
Sorry, I didn't forget this but haven't yet found the time to look into it, but will hopefully do so during the next two weeks.

**Comment by Splines** [comment]
So, finally I was able to test your new parser-level approach with ERB syntax and it works like a charm. Thank you so much for this ❤

For people stumbling on this issue: the documentation for integrating template engines with this plugin is now available [here](https://html-eslint.org/docs/integrating-template-engine).
